### PR TITLE
Refactored SDK command to use downloads.titaniumsdk.com

### LINKED ISSUE #585: Create Titanium build registry
Releases are hosted on GitHub and the Titanium CLI directly queries GitHub's API to get those releases, however we are constantly hitting the API limit. This not only affects our users, but our GitHub actions.

Furthermore, we are not able to install branch builds from the Titanium CLI including branches such as "master".

The proposed solution is comprised of 3 parts:
1. Create a new builds.titaniumsdk.com website with GitHub action to regenerate list of builds and releases
2. Wire up titanium_mobile repo to trigger regeneration of list of builds and releases
3. Update Titanium CLI to query this new builds registry

builds.titaniumsdk.com serves two purposes: a front-end for viewing builds and hosting static .json files containing the build information. The repo for this site will have a GitHub action that uses the GitHub API to query the titanium_mobile repo and regenerate static .json files for releases and branch builds. Each build in the file will contain the build hash, download URL, and timestamp. After the .json files have been generated, the site is redeployed.

We will need a new GitHub action in titanium_mobile that triggers when a release or branch build completes. It then uses the repository-dispatch action to trigger the builds site to regenerate the registry.

Lastly, we update the Titanium CLI to directly query those .json files. This is exactly how it used to work before the GitHub releases.

### PR DESCRIPTION
**Pull request by cb1kenobi**
Fixes #585

### CODE REVIEW

File: lib/commands/sdk.js [hunk 1]
```diff
@@ -369,91 +319,74 @@ SdkSubcommands.select = {
 	fn: callbackify(select)
 };
 
+function sortVersion(a, b) {
+	const [ _, amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(versionRegExp);
```
Comment (ewanharris) [inline lib/commands/sdk.js:323]: No need to have the `_` here unless you're doing it to make it obvious
```suggestion
	const [ , amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(versionRegExp);
```

File: lib/commands/sdk.js [hunk 2]
```diff
@@ -369,91 +319,74 @@ SdkSubcommands.select = {
...
+	const [ __, bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(versionRegExp);
```
Comment (ewanharris) [inline lib/commands/sdk.js:324]: Same here
```suggestion
	const [ , bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(versionRegExp);
```

File: lib/commands/sdk.js [hunk 3]
```diff
@@ -535,713 +473,575 @@ async function select(logger, config, cli) {
...
 async function install(logger, config, cli) {
-	const installLocation = afs.resolvePath(cli.env.installPath);
-
-	// make sure the install location exists
-	try {
-		await ensureInstallLocation(installLocation);
-	} catch (error) {
-		logger.error(error);
-		throw error;
+	const titaniumDir = afs.resolvePath(cli.env.installPath);
```
Comment (ewanharris) [inline lib/commands/sdk.js:536]: You've spaced the others nicely, do that here too if you're wanting to do that?

File: lib/commands/sdk.js [hunk 4]
```diff
@@ -535,713 +473,575 @@ async function select(logger, config, cli) {
...
 
-	// if (branch) {
-	// 	return doBranch(logger, config, cli, version, branch, osName);
-	// }
-	return doReleases(logger, config, cli, version, osName);
-}
+		// step 1.5: download the file
 
-/**
- * Validates the zip that has been requested to be installed. This is because when downloading a
```
Comment (ewanharris) [inline lib/commands/sdk.js:718]: I think it's good to keep the context of why we look inside the zip for some things

File: lib/commands/sdk.js [hunk 5]
```diff
@@ -1406,179 +1206,78 @@ async function uninstall(logger, config, cli) {
...
- * @property {string} ts - timestamp suffix from filename (i.e. '20200915080356')
- * @property {Date} date - date object built from timestamp in filename
- * @property {string} dateFormatted - formatted date string (i.e. '9/29/2020 11:16 AM')
+ * Retrieves the list of branches.
+ * @returns {Promise<Branches>}
  */
+async function getBranches() {
+	return Object
+		.entries((await got('https://downloads.titaniumsdk.com/registry/branches.json', { responseType: 'json' })).body)
+		.filter(([ name, count ]) => count)
```
Comment (ewanharris) [inline lib/commands/sdk.js:1266]:
```suggestion
		.filter(([, count ]) => count)
```

File: lib/commands/sdk.js [hunk 6]
```diff
@@ -1406,179 +1206,78 @@ async function uninstall(logger, config, cli) {
...
+		got('https://downloads.titaniumsdk.com/registry/ga.json', { responseType: 'json' }).then(res => ({ type: 'ga', releases: res.body }))
+	])).flatMap(r => {
+		return r.status === 'fulfilled' ? r.value.releases.map(rel => {
+			rel.type = r.value.type;
+			return rel;
+		}) : [];
+	}).filter(r => {
+		return r.assets.some(a => a.os === os);
+	}).sort((a, b) => {
+		const [ _, amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(releaseRE);
```
Comment (ewanharris) [inline lib/commands/sdk.js:1236]:
```suggestion
		const [ , amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(releaseRE);
```

File: lib/commands/sdk.js [hunk 7]
```diff
@@ -1406,179 +1206,78 @@ async function uninstall(logger, config, cli) {
...
+		const [ __, bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(releaseRE);
```
Comment (ewanharris) [inline lib/commands/sdk.js:1237]:
```suggestion
		const [ , bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(releaseRE);
```

**Review by ewanharris — COMMENTED** [review]
Some comments from a first pass, need to do a functional test

**Comment by ewanharris** [comment]
@cb1kenobi what're your plans for the existing tests for this?

**Comment by m1ga** [comment]
`ti sdk select 10.1.1.GA` (all select commands) will throw:
```
[ERROR] TypeError: Cannot read properties of undefined (reading 'toLowerCase')
    at sortVersion (/home/miga/tools/titanium/lib/commands/sdk.js:323:53)
    at Array.sort (<anonymous>)
    at Function.select (/home/miga/tools/titanium/lib/commands/sdk.js:358:4)
    at Function.selectCallbackified (node:util:304:5)
    at run (/home/miga/tools/titanium/lib/commands/sdk.js:75:29)
    at CLI.<anonymous> (/home/miga/tools/titanium/lib/cli.js:808:5)
    at wrapper (/home/miga/tools/titanium/node_modules/async/dist/async.js:271:20)
    at Object.iteratorCallback (/home/miga/tools/titanium/node_modules/async/dist/async.js:501:17)
    at Object.<anonymous> (/home/miga/tools/titanium/node_modules/async/dist/async.js:327:20)
    at CLI._fireHookCallback (/home/miga/tools/titanium/lib/hook.js:274:12)
```

a and b don't have a `name` in my tests, its only the string

**Comment by m1ga** [comment]
minor issue: 
`titanium sdk install 11.0.0.RC` will always download the file and after that says `already installed`. The current version checks it before downloading.
With `--force` it worked fine :+1:

File: lib/commands/sdk.js (hunk 1 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed.

File: lib/commands/sdk.js (hunk 3 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed.

File: lib/commands/sdk.js (hunk 5 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed.

File: lib/commands/sdk.js (hunk 6 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed.

File: lib/commands/sdk.js (hunk 7 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed.

**Comment by m1ga** [comment]
If you have a master build folder like `11.0.0` the `ti sdk select` regex will fail: https://github.com/tidev/titanium/blob/96d4881ad5a0badcb48c8936bcc152338cc84cab/lib/commands/sdk.js#L321
My `a` and `b` are: `11.1.0 9.3.2.GA` and it won't match `11.1.0`

```
[ERROR] TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))
    at sortVersion (/home/miga/tools/titanium/lib/commands/sdk.js:321:45)
    at Array.sort (<anonymous>)
    at Function.select (/home/miga/tools/titanium/lib/commands/sdk.js:356:4)
    at Function.selectCallbackified (node:util:304:5)
    at run (/home/miga/tools/titanium/lib/commands/sdk.js:75:29)
    at CLI.<anonymous> (/home/miga/tools/titanium/lib/cli.js:808:5)
    at wrapper (/home/miga/tools/titanium/node_modules/async/dist/async.js:271:20)
    at Object.iteratorCallback (/home/miga/tools/titanium/node_modules/async/dist/async.js:501:17)
    at Object.<anonymous> (/home/miga/tools/titanium/node_modules/async/dist/async.js:327:20)
    at CLI._fireHookCallback (/home/miga/tools/titanium/lib/hook.js:274:12)
```

**Comment by cb1kenobi** [comment]
@m1ga Fixed the version regexp.

**Comment by m1ga** [comment]
Regex is better now :+1:  

It still failed for me but I had some folder like `10.2.0.Fix` but I did that by hand. Old CLI didn't care, but again: it's just my SDK folder. But we could add a warning like "unsupported SDK found in SDK folder" instead of 
```
[ERROR] TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))
    at sortVersion (/home/miga/tools/titanium/lib/commands/sdk.js:321:45)
   ...
```
if there is a problem there. But nothing for this release!

---

**Tested:**
* list, select (valid and invalid, version and by list) :heavy_check_mark: 
* install valid version, latest, force, ci build  :heavy_check_mark: 
* install invalid is working and shows this error  :heavy_check_mark: 
```
[ERROR] Error: Unable to find any Titanium SDK releases or CI builds that match "9.4.2"
    at Function.install (/home/miga/tools/titanium/lib/commands/sdk.js:635:10)
```
old SDK only showed it as a "log.error" so without the sdk.js line. But I'm fine with this too. Just don't want people to think its an error in the file/function!

---

_Minor_ cosmetic issues that can be changed later (if at all):
* installing a SDK that you already have will start the download and then say: its already installed use force. When you do that it will download it again. So it would be nice (like the current CLI does it) not downloading it at all and say "use force".
* installing a CLI build will just show the spinner and no progress bar. At first I thought it was stuck. If you don't get a progress for nightly builds we could just add a `.` every two second. Might fill up the screen for slower network speeds but at least it gives the impression that it is working.
* warning for unsupported SDKs [see above]

**Comment by cb1kenobi** [comment]
@m1ga I've addressed most of your concerns: fixed version regex (again), added overwrite prompt during install, and don't show stack trace for known error messages.

**Review by caspahouzer — APPROVED** [review]
Same here. Works great. 

LGTM

File: lib/commands/sdk.js (hunk 2 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed

File: lib/commands/sdk.js [hunk 8]
```diff
@@ -262,86 +225,71 @@ async function list(logger, config, cli) {
...
-			appc.version.sort(Array.from(releases.keys())).reverse().forEach(function (r) {
-				logger.log('   ' + r.cyan + (Object.prototype.hasOwnProperty.call(sdks, r) ? ' [' + __('installed') + ']' : '') + (i++ === 0 ? ' [' + __('latest') + ']' : ''));
-			});
-			i || logger.log('   ' + __('No releases found'));
+			let i = 0;
+			for (const r of releases) {
+				logger.log(`   ${r.name.padEnd(12).cyan}\
+${Intl.DateTimeFormat('en-US', { dateStyle: 'short' }).format(new Date(r.date)).padStart(8)}\
+${humanize.filesize(r.assets.find(a => a.os === os).size, 1024, 1).toUpperCase().padStart(11)}\
+${Object.hasOwn(sdks, r) ? ' [installed]' : ''}\
```
[the remark this answers is not in the document]
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed

File: lib/commands/sdk.js [hunk 9]
```diff
@@ -175,83 +159,64 @@ async function list(logger, config, cli) {
...
-		// turn Map<string, Release> into bare object we used to report
-		const releasesObj = {};
-		if (releases) {
-			releases.forEach((value, key) => releasesObj[key] = value.url);
 		}
-		var obj = {
+
+		const obj = {
 			activeSDK,
+			branch: branchBuilds.length ? {
```
Comment (ewanharris) [inline lib/commands/sdk.js:179]: `branchBuilds` can be undefined here so need to guard against that
```suggestion
			branch: branchBuilds?.length ? {
```

**Review by ewanharris — CHANGES_REQUESTED** [review]
The output of `ti sdk list --release --output json` has changed which is going to break the editors. Either we'll need to patch those ahead of this release shipping or we could massage the output?

Comparison of the two outputs [here](https://gist.github.com/ewanharris/08e9bc0e9e7b4bb524a5bb80ebba7d5d)

**Comment by m1ga** [comment]
> The output of `ti sdk list --release --output json` has changed which is going to cause issues for the editors. Either we'll need to patch those ahead of this release shipping or we could massage the output?

Good catch @ewanharris  :+1:

File: lib/commands/sdk.js (hunk 9 above)
Comment (cb1kenobi) [reply lib/commands/sdk.js]: Fixed

**Comment by cb1kenobi** [comment]
Fixed the JSON output of the releases and branches. :)

**Comment by ewanharris** [comment]
The RC inclusion I think will be an issue too, as the [logic used currently](https://github.com/tidev/titanium-editor-commons/blob/56f3338cd0ed36fa7717039af702248e824eaa8e/src/updates/titanium/sdk.ts#L57) just handles GA.

I can update the editors to handle RC (just ignore it for now) to avoid going round on this PR again, would be best to wait a few days for that to roll out I think

**Comment by cb1kenobi** [comment]
I could add new `--beta` and `--rc` options to the list command to include those in the list. What do you think?

**Comment by ewanharris** [comment]
That would be a nice addition that would probably quicker than releasing the editors. Maybe combine under a single `--preview` flag that would toggle both rather than flags for each?

**Comment by cb1kenobi** [comment]
I don't love `--preview`. I could get behind `--unstable`. Thoughts?

**Comment by cb1kenobi** [comment]
Added the `--unstable` flag.

**Pull request merged**